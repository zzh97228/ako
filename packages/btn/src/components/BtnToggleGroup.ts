import { genGroupProps } from '@lagabu/shared';
import vue, { defineComponent, h } from 'vue';
import { useGroupProvider } from '../composables';
export default defineComponent({
  name: 'btn-toggle-group',
  props: genGroupProps(),
  setup(props, context) {
    useGroupProvider(props, context);
  },
  render() {
    return h(
      'div',
      {
        class: 'btn-toggle-group',
      },
      this.$slots.default && this.$slots.default()
    );
  },
});
