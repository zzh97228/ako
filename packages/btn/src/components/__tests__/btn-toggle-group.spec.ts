import { mount, VueWrapper } from '@vue/test-utils';
import { defineComponent } from 'vue';
import BtnToggleGroup from '../BtnToggleGroup';
import Btn from '../Btn';
describe('BtnToggleGroup.ts', () => {
  let mountFunc: (options?: object) => VueWrapper<any>;
  beforeEach(() => {
    mountFunc = (opts = {}) =>
      mount(
        defineComponent({
          props: {
            multiple: Boolean,
            disabled: Boolean,
          },
          components: {
            BtnToggleGroup,
            Btn,
          },
          data() {
            return {
              v: undefined,
            };
          },
          template: `
          <btn-toggle-group :disabled="disabled" :multiple="multiple" v-model="v">
            <btn toggleable v-for="i in 5" :key="'button' + i">button-{{ i }}</btn>
          </btn-toggle-group>`,
        }),
        opts
      );
  });
  it('should toggle single btn', async () => {
    const wrapper = mountFunc();
    const buttons = wrapper.findAllComponents({
      name: 'btn',
    });
    expect(wrapper.classes()).toContain('btn-toggle-group');
    await buttons[0].trigger('click');
    expect(buttons[0].classes()).toContain('btn--active');
    expect(wrapper.vm.v).toEqual(0);
    await buttons[1].trigger('click');
    expect(wrapper.vm.v).toEqual(1);
    expect(buttons[0].classes()).not.toContain('btn--active');
    expect(buttons[1].classes()).toContain('btn--active');
    expect(wrapper.html()).toMatchSnapshot();
    wrapper.vm.v = 2;
    await wrapper.vm.$nextTick();
    expect(buttons[0].classes()).not.toContain('btn--active');
    expect(buttons[1].classes()).not.toContain('btn--active');
    expect(buttons[2].classes()).toContain('btn--active');
  });

  it('should toggle multiple btn', async () => {
    const wrapper = mountFunc({
      props: {
        multiple: true,
      },
    });
    const buttons = wrapper.findAllComponents({
      name: 'btn',
    });
    expect(wrapper.classes()).toContain('btn-toggle-group');
    await buttons[0].trigger('click');
    expect(buttons[0].classes()).toContain('btn--active');
    expect(wrapper.vm.v).toEqual([0]);
    await buttons[1].trigger('click');
    expect(wrapper.vm.v).toEqual([0, 1]);
    expect(buttons[0].classes()).toContain('btn--active');
    expect(buttons[1].classes()).toContain('btn--active');
    expect(wrapper.html()).toMatchSnapshot();
    wrapper.vm.v = [1, 2, 3];
    await wrapper.vm.$nextTick();
    expect(buttons[0].classes()).not.toContain('btn--active');
    expect(buttons[1].classes()).toContain('btn--active');
    expect(buttons[2].classes()).toContain('btn--active');
    expect(buttons[3].classes()).toContain('btn--active');
  });

  it('should not toggle when group disabled', async () => {
    const wrapper = mountFunc({
      props: {
        disabled: true,
      },
    });

    const buttons = wrapper.findAllComponents({
      name: 'btn',
    });
    expect(wrapper.classes()).toContain('btn-toggle-group');
    await buttons[0].trigger('click');
    expect(buttons[0].classes()).not.toContain('btn--active');
    wrapper.vm.v = 0;
    await wrapper.vm.$nextTick();
    expect(buttons[0].classes()).not.toContain('btn--active');
  });
});
