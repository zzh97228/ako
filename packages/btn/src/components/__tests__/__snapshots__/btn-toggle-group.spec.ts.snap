// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BtnToggleGroup.ts should toggle multiple btn 1`] = `
<div class="btn-toggle-group">
  <button type="button"
          class="btn elevation-xs btn--active"
  >
    <div class="btn__content">
      button-1
    </div>
  </button>
  <button type="button"
          class="btn elevation-xs btn--active"
  >
    <div class="btn__content">
      button-2
    </div>
  </button>
  <button type="button"
          class="btn elevation-xs"
  >
    <div class="btn__content">
      button-3
    </div>
  </button>
  <button type="button"
          class="btn elevation-xs"
  >
    <div class="btn__content">
      button-4
    </div>
  </button>
  <button type="button"
          class="btn elevation-xs"
  >
    <div class="btn__content">
      button-5
    </div>
  </button>
</div>
`;

exports[`BtnToggleGroup.ts should toggle single btn 1`] = `
<div class="btn-toggle-group">
  <button type="button"
          class="btn elevation-xs"
  >
    <div class="btn__content">
      button-1
    </div>
  </button>
  <button type="button"
          class="btn elevation-xs btn--active"
  >
    <div class="btn__content">
      button-2
    </div>
  </button>
  <button type="button"
          class="btn elevation-xs"
  >
    <div class="btn__content">
      button-3
    </div>
  </button>
  <button type="button"
          class="btn elevation-xs"
  >
    <div class="btn__content">
      button-4
    </div>
  </button>
  <button type="button"
          class="btn elevation-xs"
  >
    <div class="btn__content">
      button-5
    </div>
  </button>
</div>
`;
